### PR TITLE
Fix crash in lightbulbs in projection scenarios.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
A previous change of mine moved it so that we try to grab the user's selectoin as the first
thing we do when queried in lightbulbs.  This was to make it so we would not call back into
the UI thread from a BG thread.  Unfortunately, the code i moved was not safe to call if the
query was a projection buffer.  This wasn't a problem previously because there was a prior
check that ran that would cause us to bail out.

I've now reordered things so that the check comes first, and then we get the selection afterwards.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide @jasonmalinowski

**Comment by CyrusNajmabadi** [comment]
Tagging @MattGertz 


**Customer scenario**

Typing in a projection file, with an inert section, will crash when the editor calls into us to see if there are lightbulbs.

**Bugs this fixes:**

nonw.

**Workarounds, if any**

none.

**Risk**

Low

**Performance impact**

None.

**Is this a regression from a previous update?**

Yes.

**Root cause analysis:**

This was due to a recent change i made to prevent a deadlock with the editor with the new work they're doing with lightbulbs for us.   We actually have a test for this, but due to problems we were having with our EditorTestApp tests not running, this slipped through. 

This change actually gets us back to a green state with integration tests so we can depend on them for catching bugs like this.

**How was the bug found?**

Integration tests.

**Comment by MattGertz** [comment]
Approved pending tests/CRs.

File: src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSource.cs [hunk 1]
```diff
@@ -640,8 +640,18 @@ private static SuggestedActionSetPriority GetSuggestedActionSetPriority(CodeActi
...
+                    // we are called with a range snapshot that refers to the projection buffer
+                    // and not the actual roslyn code that is being projected into it.
+                    var document = range.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+                    if (document == null)
+                    {
+                        return false;
+                    }
+
+                    // Next, before we do any async work, acquire the user's selection, directly grabbing
+                    // it from teh UI thread if htat's what we're on. That way we don't have any reentrancy
```
Comment (Pilchie) [inline src/EditorFeatures/Core/Implementation/Suggestions/SuggestedActionsSource.cs]: Spelling: `teh`

**Comment by jasonmalinowski** [comment]
I agree @Pilchie's question seems like a good one: I would expect the editor to be doing mapping. But if this unblocks our test failures, let's get this in first.

**Comment by CyrusNajmabadi** [comment]
@Pilchie I'm surprised as well.  And i'm not at all certain that that just isn't a bug on the editor side.  However, right now that's what they're doing (and they've likely been doing it for ages), so this just reverts us to the behavior where we gracefully handle it :)

**Comment by Pilchie** [comment]
Looping in @olegtk here.

**Comment by olegtk** [comment]
Just looking at our code we call HasSuggestedActionsAsync() for both projection and mapped subject buffers (obviously filtering providers that match those buffers by content type). That's have it has been since Dev14.

**Comment by Pilchie** [comment]
But we don't export ours for projection content types.  From [here](http://source.roslyn.io/#Microsoft.CodeAnalysis.EditorFeatures/Implementation/Suggestions/SuggestedActionsSourceProvider.cs,21):

``` C#
    [Export(typeof(ISuggestedActionsSourceProvider))]
    [Export(typeof(SuggestedActionsSourceProvider))]
    [VisualStudio.Utilities.ContentType(ContentTypeNames.RoslynContentType)]
    [VisualStudio.Utilities.ContentType(ContentTypeNames.XamlContentType)]
    [VisualStudio.Utilities.Name("Roslyn Code Fix")]
    [VisualStudio.Utilities.Order]
    internal partial class SuggestedActionsSourceProvider : ISuggestedActionsSourceProvider
```

**Comment by olegtk** [comment]
weird. Is it a regression?

**Comment by Pilchie** [comment]
Not sure - we had that code to filter out null documents for a long time.  It could be someone added it to work around a longstanding bug.

**Comment by olegtk** [comment]
probably long standing bug, please file it.

**Comment by Pilchie** [comment]
@CyrusNajmabadi - did you get an internal bug filed for this?

**Comment by CyrusNajmabadi** [comment]
No.

**Comment by CyrusNajmabadi** [comment]
Here you go: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/428944

**Pull request merged**